### PR TITLE
Feat/consider short term schedule [WIP]

### PR DESCRIPTION
**Pull request by L-M-Sherlock**
| Algorithm | Log Loss | RMSE (bins) | Parameters |
| --- | --- | --- | --- |
| **FSRS-4.5** | 0.3236±0.1512 | 0.0433±0.0310 | 17 |
| FSRS-4.5 + short-term | 0.3206±0.1489 | 0.0425±0.0302 | 17 |
| FSRS-4.5 + short-term + w[17] | 0.3204±0.1487 | 0.0424±0.0299 | 18 |
| FSRS-4.5 + short-term + w[17] + w[18] | 0.3195±0.1483 | 0.0415±0.0297 | 19 |
| FSRS-4.5 + short-term + w[17] + w[18] + w[19] | 0.3194±0.1482 | 0.0415±0.0298 | 20 |

| Algorithm                                     | Log Loss      | RMSE (bins)   | Parameters |
| --------------------------------------------- | ------------- | ------------- | ---------- |
| **FSRS-4.5**                                  | 0.3461±0.1681 | 0.0628±0.0455 | 17         |
| FSRS-4.5 + short-term                         | 0.3417±0.1650 | 0.0608±0.0442 | 17         |
| FSRS-4.5 + short-term + w[17]                 | 0.3418±0.1650 | 0.0608±0.0439 | 18         |
| FSRS-4.5 + short-term + w[17] + w[18]         | 0.3411±0.1648 | 0.0600±0.0439 | 19         |
| FSRS-4.5 + short-term + w[17] + w[18] + w[19] | 0.3410±0.1648 | 0.0600±0.0439 | 20         |

Related code:

```python
    def stability_short_term(self, state: Tensor, rating: Tensor) -> Tensor:
        new_s = state[:, 0] * torch.exp(self.w[17] * (rating - 3 + self.w[18]))
        return new_s

    def next_d_short_term(self, state: Tensor, rating: Tensor) -> Tensor:
        new_d = state[:, 1] - self.w[19] * (rating - 3)
        return new_d
```

@Expertium, I finished the benchmarking of four candidates considering short-term reviews. I prefer the `FSRS-4.5 + short-term` because it doesn't introduce any extra parameters and reduce the RMSE(bins) 3.2% (weighted by ln(number of reviews)). What do you think of?

### CODE REVIEW

**Comment by Expertium** [comment]
I suppose we can release it in the future. Btw, if you include it in the table, it will be a little misleading. Right now it says 708,151,820 reviews, but this algorithm uses more reviews. You should add some sort of clarification regarding that.

**Comment by Expertium** [comment]
Alternatively, we could use the version with new parameters and call it FSRS v5, or at least alpha v5.

**Comment by L-M-Sherlock** [comment]
I plan to release FSRS-5 in 2024H1. So it's not urgent. Many improvement ideas are welcome.

**Comment by Expertium** [comment]
Well, my new definition of D improves RMSE by 2%, but it adds a lot of new parameters and is incompatible with short-term scheduling since R=100% for same-day reviews.
One interesting idea that I have is doing what ACT-R is doing, and scaling time for short-term reviews, like this: forgetting_curve(S, h*delta_t). This could solve the R=100% problem.

**Comment by Expertium** [comment]
@L-M-Sherlock So here's my idea for short-term scheduling:
1) Add a new parameter that scales real time to convert it into "psychological" time. It should be equal to 1 for reviews that didn't happen on the same day, and >1 for same-day reviews. Note that I'm assuming that delta_t is expressed as days. For example, 5 minutes would be 5/3600=0.001388889.
2) The clampings should be `w[18]=w[18].clamp(2, 3600)`. If it's 2, then 12 hours is converted into 1 day. If it's 3600, then 1 minute is converted into 1 day. The default should be 240, which would convert 15 minutes into one day.
3) R should be calculated like this:
```
time_scaling = torch.where(X[:, 0] < 1, self.w[18], 1)
r = power_forgetting_curve(X[:, 0] * time_scaling , state[:, 0])
```
X[:, 0] will have to be a float, of course.

Now you can use the same formulas for both long- and short-term scheduling.

**Comment by Expertium** [comment]
@L-M-Sherlock once you have some free time, I would like you to benchmark my suggestion in the comment above. Maybe with one minor change:
```
time_scaling = torch.where(X[:, 0] < 1, np.exp(self.w[18]), 1)
r = power_forgetting_curve(X[:, 0] * time_scaling , state[:, 0])
w[18]=w[18].clamp(np.log(2), np.log(3600))
```
Hopefully, this should make it easier to optimize.

**Comment by L-M-Sherlock** [comment]
It requires to re-build the dataset because I haven't included the float interval in it.

**Comment by L-M-Sherlock** [comment]
Weighted average by reviews:

| Algorithm | Log Loss | RMSE (bins) | Parameters |
| --- | --- | --- | --- |
| FSRS-4.5 | 0.3252±0.1514 | 0.0533±0.0334 | 17 |
| FSRS-4.5 + short-term | 0.3214±0.1489 | 0.0520±0.0331 | 19 |

Weighted average by log(reviews):

| Algorithm | Log Loss | RMSE (bins)   | Parameters |
| --- | --- | --- | --- |
| FSRS-4.5 | 0.3485±0.1701 |  0.0733±0.0474 | 17 |
| FSRS-4.5 + short-term | 0.3439±0.1673 | 0.0715±0.0467 | 19 |

**Comment by Expertium** [comment]
Maybe add more parameters? If you're planning to include this in the next major version of FSRS, then why not?

**Comment by L-M-Sherlock** [comment]
> Maybe add more parameters? If you're planning to include this in the next major version of FSRS, then why not?

I have tried to add more parameters into the short-term module, but the improvement is less than 0.5%.

**Comment by L-M-Sherlock** [comment]
Weighted average by reviews:

| Algorithm | Log Loss | RMSE (bins) | Parameters |
| --- | --- | --- | --- |
| FSRS-4.5 | 0.3252±0.1514 | 0.0533±0.0334 | 17 |
| FSRS-4.5 + short-term | 0.3212±0.1489 | 0.0517±0.0328 | 19 |

Weighted average by log(reviews):

| Algorithm | Log Loss | RMSE (bins) | Parameters |
| --- | --- | --- | --- |
| FSRS-4.5 | 0.3485±0.1701 |  0.0733±0.0474 | 17 |
| FSRS-4.5 + short-term | 0.3430±0.1670 | 0.0705±0.0460 | 19 |

Good News: After tuning the default parameters, considering short-term reduces RMSE(bins) 3%.
